zcash_client_backend/src/proto.rs

```
@@ -261,6 +262,85 @@ impl<SpendAuth> From<&orchard::Action<SpendAuth>> for compact_formats::CompactOr
+    /// representable within a `u32`.
+    pub fn sapling_activation_height(&self) -> BlockHeight {
+        self.sapling_activation_height
+            .try_into()
+            .expect("lightwalletd should provide in-range heights")
+    }
+
+    /// Returns the current consensus branch ID for the chain tip of the chain this server
+    /// is following, or `None` if it is not recognised.
+    pub fn consensus_branch_id(&self) -> Option<consensus::BranchId> {
+        u32::from_str_radix(&self.consensus_branch_id, 16)
+            .ok()?
+            .try_into()
+            .ok()
+    }
+
+    /// Returns the chain tip height reported by the full node backing this server.
+    ///
+    /// If the full node is still syncing, this may not be the network's chain tip; in
+    /// this case, [`Self::estimated_height`] will report a larger height.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `LightdInfo.block_height` is not representable within
+    /// a `u32`.
+    pub fn block_height(&self) -> BlockHeight {
+        self.block_height
+            .try_into()
+            .expect("lightwalletd should provide in-range heights")
+    }
+
+    /// Returns the estimated chain tip height for the chain this server is following.
+    ///
+    /// If the full node backing this server is fully synced, this is always equal to
+    /// [`Self::block_height`].
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `LightdInfo.estimated_height` is not representable
+    /// within a `u32`.
+    pub fn estimated_height(&self) -> BlockHeight {
+        self.estimated_height
+            .try_into()
+            .expect("lightwalletd should provide in-range heights")
+    }
+
+    /// Returns the donation address for this server.
+    ///
+    /// Returns `None` if:
+    /// - no donation address was provided.
+    /// - the provided donation address is not a valid [`unified::Address`].
```

Thanks for adding this. It would be nice to have an Error state rather than just "None" - silently failing when someone provides an invalid address in their config feels like bad CLI UX.

---

This is not CLI UX. This is the API that wallet libraries use; if the server configured an invalid address, there is nothing the wallet can do.

---

If a server operator launches Zaino with an invalid UA in their config file I want Zaino to crash with an informative error. How can we do that if this method returns the same as an empty address? That's what I meant by "CLI UX" but was admittedly too vague of a term.

---

Zaino should be checking the value of this before putting it into a gRPC response, ideally by just using the `unified::Address` type in its in-memory representation. It would then error at that point which is far more helpful than hoping the operator double-checks the gRPC output after deploying their config into production.